streaming/src/java/org/apache/hive/streaming/HiveStreamingConnection.java

```diff
@@ -496,21 +560,46 @@ private void checkState() throws StreamingException {
     if (currentTransactionBatch == null) {
       throw new StreamingException("Transaction batch is null. Missing beginTransaction?");
     }
-    if (currentTransactionBatch.state != TxnState.OPEN) {
+    if (currentTransactionBatch.getCurrentTransactionState() != TxnState.OPEN) {
       throw new StreamingException("Transaction state is not OPEN. Missing beginTransaction?");
     }
   }
 
   @Override
   public void beginTransaction() throws StreamingException {
     checkClosedState();
+    partitions = new HashSet<>();
     beginNextTransaction();
   }
 
   @Override
   public void commitTransaction() throws StreamingException {
+    commitTransactionWithPartition(null);
+  }
+
+  @Override
+  public void commitTransactionWithPartition(Set<String> partitions)
+      throws StreamingException {
     checkState();
-    currentTransactionBatch.commit();
+
+    Set<String> createdPartitions = new HashSet<>();
+    if (partitions != null) {
+      for (String partition: partitions) {
+        try {
+          PartitionInfo info = createPartitionIfNotExists(
+              Warehouse.getPartValuesFromPartName(partition));
+          if (!info.isExists()) {
+            createdPartitions.add(partition);
+          }
+        } catch (MetaException e) {
+          throw new StreamingException("Partition " + partition + " is invalid.", e);
+        }
+      }
+    }
+
+    currentTransactionBatch.commitWithPartitions(createdPartitions);
+    this.partitions.addAll(
+        currentTransactionBatch.getPartitions());
     connectionStats.incrementCommittedTransactions();
```

Comment:
log total added partitions? now that we have central place (driver) that commits these partitions, can you add total-partitions and created-partitions to connectionStats for debugging?